engine/src/main/scala/cromwell/engine/workflow/lifecycle/finalization/WorkflowCallbackActor.scala

```diff
@@ -0,0 +1,226 @@
+      // one there, do not perform a callback.
+      val callbackUri: Option[URI] = requestedCallbackUri.map(WorkflowCallbackConfig.createAndValidateUri).getOrElse(defaultCallbackUri)
+      callbackUri.map { uri =>
+        performCallback(workflowId, uri, terminalState, outputs, failures) onComplete {
+          case Success(_) =>
+            log.info(s"Successfully sent callback for workflow for workflow $workflowId in state $terminalState to $uri")
+            sendMetadata(workflowId, successful = true, uri)
+          case Failure(t) =>
+            log.warning(s"Permanently failed to send callback for workflow $workflowId in state $terminalState to $uri: ${t.getMessage}")
+            sendMetadata(workflowId, successful = false, uri)
+        }
+      }.getOrElse(())
+    case Broadcast(ShutdownCommand) | ShutdownCommand => context stop self
+    case other => log.warning(s"WorkflowCallbackActor received an unexpected message: $other")
+  }
+
+  private def makeHeaders: Future[List[HttpHeader]] = {
+    authMethod.toList.map(_.getAccessToken).map {
+      case Valid(header) => Future.successful(header)
+      case Invalid(err) => Future.failed(new RuntimeException(err.toString)) // TODO better error
+    }
+      .map(t => t.map(t => RawHeader("Authorization", s"Bearer $t")))
+      .traverse(identity)
+  }
+
+  private def performCallback(workflowId: WorkflowId, callbackUri: URI, terminalState: WorkflowState, outputs: CallOutputs, failures: List[String]): Future[Unit] = {
+    // Only send outputs if the workflow succeeded
+    val callbackPostBody = terminalState match {
+      case WorkflowSucceeded => CallbackMessage(workflowId.toString, terminalState.toString, Option(outputs.outputs.map(entry => (entry._1.name, entry._2))), None)
+      case WorkflowFailed => CallbackMessage(workflowId.toString, terminalState.toString, None, Option(failures))
+      case _ => CallbackMessage(workflowId.toString, terminalState.toString, None, None)
+    }
+    for {
+      entity <- Marshal(callbackPostBody).to[RequestEntity]
+      headers <- makeHeaders
+      request = HttpRequest(method = HttpMethods.POST, uri = callbackUri.toString, entity = entity).withHeaders(headers)
+      response <- withRetry(
+        () => sendRequestOrFail(request),
+        backoff = backoffWithDefault,
+        maxRetries = Option(maxRetriesWithDefault),
+        onRetry = err => log.warning(s"Will retry after failure to send workflow callback for workflow $workflowId in state $terminalState to $callbackUri : $err")
+      )
+      result <- {
+        response.entity.discardBytes()
+        Future.unit
+      }
+    } yield result
+  }
+
+  private def sendRequestOrFail(request: HttpRequest): Future[HttpResponse] =
+    httpClient.sendRequest(request).flatMap(response =>
+      if (response.status.isFailure()) {
+        response.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String) flatMap { errorBody =>
```

Comment:
For a scheduler with tons of IO, I feel Cromwell has under-invested in creating a standard util collection of HTTP operations; I recognize this line from the TES backend. Creating such a one is highly optional in the context of this PR, of course.

Reply:
Totally agree that we should have this but unfortunately don't have time to add it here. 😞